source/MaterialXGenShader/HwShader.cpp

```diff
@@ -127,4 +127,18 @@ void HwShader::getTopLevelShaderGraphs(ShaderGenerator& shadergen, std::deque<Sg
     }
 }
 
+bool HwShader::hasSamplerNodes() const
+{
+    bool sampling = false;
+    for (SgNode* node : getNodeGraph()->getNodes())
+    {
+        if (node->hasClassification(SgNode::Classification::CONVOLUTION2D))
+        {
+            sampling = true;
+            break;
+        }
+    }
+    return sampling;
+}
```

Comment:
For performance, instead of this linear search, the graph can be classified as convolution as soon as a convolution node is added during creation. The graph class is derived from SgNode so graphs have the same support for classifications as atomic nodes.

Then there is already a method on the Shader class to query classifications: Shader::hasClassification(), which in turn calls hasClassification on the shader graph. So instead of calling shader->hasSamplerNodes() we can then call shader->hasClassification(SgNode::Classification::CONVOLUTION2D)

Reply:
I didn't know this so wasn't sure where to cache. I'll make the change.